File: 100/116 - Populating Next Right Pointers in Each Node.c
/**
 * Definition for binary tree with next pointer.
 * struct TreeLinkNode {
 *  int val;
 *  struct TreeLinkNode *left, *right, *next;
 * };
 *
 */
void connect(struct TreeLinkNode *root) {
    struct TreeLinkNode *first = root;

    while (first) {
        root = first;
        first = first->left;
        
        while (root && root->left) {
            root->left->next = root->right;
            root->right->next = root->next ? root->next->left : NULL;
            root = root->next;
        }
    }
}

File: csharp/116.populating-next-right-pointers-in-each-node.cs
/*
 * @lc app=leetcode id=116 lang=csharp
 *
 * [116] Populating Next Right Pointers in Each Node
 */
/*
// Definition for a Node.
public class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node(){}
    public Node(int _val,Node _left,Node _right,Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
}
*/
public class Solution {
    public Node Connect(Node root) {
        Node first = root;
        while (first != null) {
            Node curr = first;
            while (curr != null && curr.left != null) {
                curr.left.next = curr.right;
                curr.right.next = curr.next == null ? null : curr.next.left;
                curr = curr.next;
            }
            first = first.left;
        }
        return root;
    }
}

